Allow selection of invalid geometries

nyalldawson opened the pull request:
When selecting features via rect, ignore GEOS errors due to invalid geometries and allow these features to be selected based on their  bounding boxes alone.

Fixes #38460

I'm not 100% on this fix. I can't see why this behavior has changed since 3.10, unless it's caused by changes in GEOS itself. Feedback is appreciated!!

nyalldawson committed:
When selecting features via rect, ignore GEOS errors due to invalid
geometries and allow these features to be selected based on their
bounding boxes alone.

Fixes #38460

uclaros commented:
This will select false positives, ie. invalid geometries that may not even be visible on the current extent but their mbr is. There will be no visual clue that this has happened, apart from the number of selected features.
The current problematic behavior though is falsely not selecting invalid geometries, which has the selection highlighting (or lack of) as an obvious visual clue.

nyalldawson commented:
@uclaros actually the situation is a bit more complex then that. The iterator is being requested with the Exact Intersection flag here, so we already can be confident that they do have more then just a bounding box intersection (and if not then it's a bug in the particular data provider). I suspect that the test in the non-contains branch here is actually redundant and could be removed.

uclaros commented:
I think that's true only for the rectangle selection where the rectangle is the same as its mbr. For polygonal selection though, the exact intersection does not suffice and the non-contains branch is also required.

nyalldawson commented:
Good point! How do you think we should handle this then?

uclaros commented:
I don't know, maybe something like this?
```
if ( !selectGeomTrans.contains( g ) &&
     !selectGeomTrans.lastError().isEmpty() &&
     !selectGeomTrans.contains( g.makeValid() ) )
  continue;
```
I think this will give a second chance to the invalid geometry.
But, is `lastError` reset after a successful `contains()`?
And what if the `selectGeomTrans` is invalid instead?

nyalldawson commented:
> I don't know, maybe something like this?

Yep, that should work!

> But, is lastError reset after a successful contains()?

Yep, it's reset before calling any QgsGeometry method which can potentially set it.

> And what if the selectGeomTrans is invalid instead?

We could just fix the selection geom once before testing any features -- e.g. buffer it by 0.

uclaros commented:
We could also check if `selectGeometry` is the same as its mbr, ie selection was made by rectangle. In that case, completely skip the redundant non-contains branch. This I believe is the most used case (the default select by rectangle without pressing alt) so avoiding the extra intersection check can only be good.

Related but slightly off topic thought: would it make sense to move the refined selection logic to `QgsFeatureRequest` by adding some kind of `setFilterGeometry()` method that would support all spatial predicates? This logic is probably repeated in various other places and could be centralized.

nyalldawson commented:
> Related but slightly off topic thought: would it make sense to move the refined selection logic to QgsFeatureRequest by adding some kind of setFilterGeometry() method that would support all spatial predicates? This logic is probably repeated in various other places and could be centralized.

That'd be great! (It's not a trivial piece of work though...)

nyalldawson committed:
Safer approach to handling invalid geometries in selection tools

nyalldawson committed:
Add method to silence noisy GEOS error logging handling when not
desirable.

By default these errors are logged to the console and in the QGIS UI.
But for some operations errors are expected and logging these just results
in noise.

nyalldawson committed:
Silence noisy GEOS messages when selecting invalid geometries, fix logic

nyalldawson committed:
Use prepared geometries for faster selections

nyalldawson commented:
@uclaros (much) improved version pushed now.

uclaros commented:
That's (much) better!
Maybe we should also log a failed second chance in case `makeValid()` fails?

nyalldawson committed:
Log message if selection cannot be determined even after geometry repair

nyalldawson commented:
> Maybe we should also log a failed second chance in case makeValid() fails?

Done!
